config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java

```diff
@@ -0,0 +1,228 @@
+import org.springframework.security.oauth2.server.resource.web.BearerTokenAuthenticationFilter;
+import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver;
+import org.springframework.security.oauth2.server.resource.web.DefaultBearerTokenResolver;
+import org.springframework.security.oauth2.server.resource.web.access.BearerTokenAccessDeniedHandler;
+import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+
+/**
+ *
+ * An {@link AbstractHttpConfigurer} for OAuth 2.0 Resource Server Support.
+ *
+ * By default, this wires a {@link BearerTokenAuthenticationFilter}, which can be used to parse the request
+ * for bearer tokens and make an authentication attempt.
+ *
+ * <p>
+ * The following configuration options are available:
+ *
+ * <ul>
+ * <li>{@link #jwt()} - enables Jwt-encoded bearer token support</li>
+ * </ul>
+ *
+ * <p>
+ * When using {@link #jwt()}, a Jwk Set Uri must be supplied via {@link JwtConfigurer#jwkSetUri}
+ *
+ * <h2>Security Filters</h2>
+ *
+ * The following {@code Filter}s are populated when {@link #jwt()} is configured:
+ *
+ * <ul>
+ * <li>{@link BearerTokenAuthenticationFilter}</li>
+ * </ul>
+ *
+ * <h2>Shared Objects Created</h2>
+ *
+ * The following shared objects are populated:
+ *
+ * <ul>
+ * <li>{@link SessionCreationPolicy} (optional)</li>
+ * </ul>
+ *
+ * <h2>Shared Objects Used</h2>
+ *
+ * The following shared objects are used:
+ *
+ * <ul>
+ * <li>{@link AuthenticationManager}</li>
+ * </ul>
+ *
+ * If {@link #jwt()} isn't supplied, then the {@link BearerTokenAuthenticationFilter} is still added, but without
+ * any OAuth 2.0 {@link AuthenticationProvider}s. This is useful if needing to switch out Spring Security's Jwt support
+ * for a custom one.
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ * @see BearerTokenAuthenticationFilter
+ * @see JwtAuthenticationProvider
+ * @see NimbusJwtDecoderJwkSupport
+ * @see AbstractHttpConfigurer
+ */
+public final class OAuth2ResourceServerConfigurer<H extends HttpSecurityBuilder<H>> extends
+		AbstractHttpConfigurer<OAuth2ResourceServerConfigurer<H>, H> {
+
+	private BearerTokenResolver resolver = new DefaultBearerTokenResolver();
+	private BearerTokenRequestMatcher matcher = new BearerTokenRequestMatcher();
+
+	private BearerTokenAuthenticationEntryPoint entryPoint = new BearerTokenAuthenticationEntryPoint();
+	private BearerTokenAccessDeniedHandler deniedHandler = new BearerTokenAccessDeniedHandler();
+
+	private JwtConfigurer jwtConfigurer;
+
+	public JwtConfigurer jwt() {
+		if (this.jwtConfigurer == null) {
+			this.jwtConfigurer = new JwtConfigurer();
+		}
+
+		return this.jwtConfigurer;
+	}
+
+	@Override
+	public void setBuilder(H http) {
+		super.setBuilder(http);
+		initSessionCreationPolicy(http);
```

Comment:
Actually, no, it can't. I can give you a brief explanation here, and ping me if it still isn't clear why.

`SessionManagementConfigurer` makes its decisions about which `SecurityContextRespository` and `RequestCache` to use inside its `init` method, and it does so based on what the user has set for `sessionCreationPolicy`. By the time this configurer sees this decision, it cannot tell whether the user made that decision or `SessionManagementConfigurer` by default.

This is one that @rwinch and I chatted about for a bit early on. I'm open to suggestions if there is a better way (make sure to see the changes I made to `SessionManagementConfigurer` and how those changes affect its `init` method, not its `configure` method.).